File: matlab/bisection_search_method.m
clear all
clc
tic 

% Initial Values
x_lower = 1;
x_upper = 2;
eps = 100;
iter = 0; % iteration number

% Bisection Algorithm
func = @(x) x^2 + 4*cos(x);   
d_func = @(x) 2*x - 4*sin(x); % derivative of function
while (eps > 10^-5)
    
    x_mid = (x_lower + x_upper)/2; 
    
    if (d_func(x_mid) > 0)
        x_upper = x_mid; 
    elseif(d_func(x_mid) < 0)
        x_lower = x_mid;
    end
    
    iter = iter + 1; 
    eps = x_upper - x_lower; 
end

timeElapsed = toc;

fprintf("\tBisection Search Method\n")
fprintf("Iteration : %d\n",iter);
fprintf("Computation time : %f\n",timeElapsed);
fprintf("Minimum point of function : x = %f, f(x) = %f\n",x_mid,func(x_mid));

File: c++/bisection_search_method.cpp
#include <bits/stdc++.h>

#define MAX 99999


double func(double x){
    // Function to minimize, this can be changed to any function.
    // Parameters:
    //      :x: variable of function
    return x * x + 4 * cos(x);   
}


double dfunc(double x){
    // Derivative of minimized function
    return 2 * x - 4 * sin(x);
}


double bisection(double xLower, double xUpper){
    // Bisection search method to find minimum of function
    // Parameters:
    //      :xUpper: initial value of upper limit
    //      :xLower: initial value of lower limit

    std::clock_t start;
    start = std::clock();

    int itr = 0; 
    double xMid, error = MAX;
    while(error > exp10(-5)){
        xMid = (xUpper + xLower) / 2;

        if (dfunc(xMid) > 0){
            xUpper = xMid;
        }
        else{
            xLower = xMid;
        }
        itr++;
        error = fabs(xUpper - xLower);
    } 

    std::cout << "Bisection Search Method" << std::endl;
    std::cout << "Time : " << (std::clock() - start) / (double)CLOCKS_PER_SEC << std::endl;
    std::cout << "Iteration : " << itr << std::endl;
    std::cout << "Minimum point of function : " << "x = " << xMid << ", f(x) = " << func(xMid) << std::endl;

    return xMid;
}


int main(){
    double xLower = 1;
    double xUpper = 2;
    bisection(xLower, xUpper);

    return 0;
}